cargo-optee/src/main.rs

```diff
@@ -0,0 +1,567 @@
+use std::process;
+
+mod ca_builder;
+mod common;
+mod config;
+mod ta_builder;
+
+use common::Arch;
+
+/// Parse environment variable in KEY=VALUE format
+fn parse_env_var(s: &str) -> Result<(String, String), String> {
+    if let Some(eq_pos) = s.find('=') {
+        let (key, value) = s.split_at(eq_pos);
+        let value = &value[1..]; // Skip the '=' character
+        Ok((key.to_string(), value.to_string()))
+    } else {
+        Err(format!(
+            "Invalid environment variable format: '{}'. Expected 'KEY=VALUE'",
+            s
+        ))
+    }
+}
+
+/// Resolve a potentially relative path to an absolute path based on the project directory
+fn resolve_path_relative_to_project(path: &PathBuf, project_path: &std::path::Path) -> PathBuf {
+    if path.is_absolute() {
+        path.clone()
+    } else {
+        project_path.join(path)
+    }
+}
+
+/// Execute TA build or install (shared logic)
+fn execute_ta_command(
+    common: CommonBuildArgs,
+    std: Option<bool>,
+    ta_dev_kit_dir: Option<PathBuf>,
+    signing_key: Option<PathBuf>,
+    uuid_path: Option<PathBuf>,
+    install_target_dir: Option<&PathBuf>,
+) -> anyhow::Result<()> {
+    // Resolve project path from manifest or current directory
+    let project_path = if let Some(manifest) = common.manifest_path {
+        let parent = manifest
+            .parent()
+            .ok_or_else(|| anyhow::anyhow!("Invalid manifest path"))?;
+
+        // Normalize: if parent is empty (e.g., manifest is just "Cargo.toml"),
+        // use current directory instead
+        if parent.as_os_str().is_empty() {
+            std::env::current_dir()?
+        } else {
+            // Canonicalize must succeed, otherwise treat as invalid manifest path
+            parent
+                .canonicalize()
+                .map_err(|_| anyhow::anyhow!("Invalid manifest path"))?
+        }
+    } else {
+        std::env::current_dir()?
+    };
+
+    // Resolve build configuration with priority: CLI > metadata > error
+    let build_config = config::BuildConfig::resolve(
+        &project_path,
+        "ta", // Component type for TA
+        common.arch,
+        Some(common.debug),
+        std, // None means read from config, Some(true/false) means CLI override
+        ta_dev_kit_dir,
+        None, // optee_client_export not needed for TA
+        signing_key,
+        uuid_path.clone(),
+    )?;
+
+    // Print the final configuration being used
+    build_config.print_config("ta", &project_path);
+
+    // Get required ta_dev_kit_dir and resolve relative to project
+    let ta_dev_kit_dir_config = build_config.require_ta_dev_kit_dir()?;
+    let ta_dev_kit_dir = resolve_path_relative_to_project(&ta_dev_kit_dir_config, &project_path);
+
+    // Validate that ta_dev_kit_dir exists (print absolute path)
+    if !ta_dev_kit_dir.exists() {
+        bail!(
+            "TA development kit directory does not exist: {:?}",
+            ta_dev_kit_dir
+        );
+    }
```

Comment:
Move this check into resolve_path_relative_to_project() to reduce the duplicate code, the same case is in line 115-118, please also check and fix other similar cases